diff --git a/FourQ_64bit_and_portable/makefile b/FourQ_64bit_and_portable/makefile
index 7b6fd07..7687e05 100644
--- a/FourQ_64bit_and_portable/makefile
+++ b/FourQ_64bit_and_portable/makefile
@@ -77,7 +77,7 @@ else ifeq "$(ARCH)" "ARM"
 endif
 endif
 
-ADDITIONAL_SETTINGS=-fwrapv -fomit-frame-pointer -march=native
+ADDITIONAL_SETTINGS=-fwrapv -fomit-frame-pointer -O3 -fPIC
 ifeq "$(EXTENDED_SET)" "FALSE"
     ADDITIONAL_SETTINGS=
 endif
@@ -115,6 +115,12 @@ OBJECTS_ALL=$(OBJECTS) $(OBJECTS_FP_TEST) $(OBJECTS_ECC_TEST) $(OBJECTS_CRYPTO_T
 
 all: crypto_test ecc_test fp_test $(SHARED_LIB_O)
 
+install: libfourq.a 
+	cp libfourq.a $(PREFIX)/lib
+	cp FourQ.h $(PREFIX)/include
+	cp FourQ_api.h $(PREFIX)/include
+	cp FourQ_internal.h $(PREFIX)/include
+
 ifeq "$(SHARED_LIB)" "TRUE"
     $(SHARED_LIB_O): $(OBJECTS)
 	    $(CC) -shared -o $(SHARED_LIB_O) $(OBJECTS)
@@ -181,6 +187,9 @@ ecc_tests.o: tests/ecc_tests.c
 fp_tests.o: tests/fp_tests.c
 	$(CC) $(CFLAGS) tests/fp_tests.c
 
+libfourq.a: $(OBJECTS)
+	$(AR) rcs libfourq.a  $(OBJECTS)
+
 .PHONY: clean
 
 clean:
diff --git a/FourQ_64bit_and_portable/schnorrq.c b/FourQ_64bit_and_portable/schnorrq.c
index da89f86..099f4eb 100644
--- a/FourQ_64bit_and_portable/schnorrq.c
+++ b/FourQ_64bit_and_portable/schnorrq.c
@@ -14,7 +14,7 @@
 #include "FourQ_params.h"
 #include "../random/random.h"
 #include "../sha512/sha512.h"
-#include <malloc.h>
+#include <stdlib.h>
 #include <string.h>
 

