diff --git a/third_party/BUILD b/third_party/BUILD
index 77cb52d0fc..c4b647f5c9 100644
--- a/third_party/BUILD
+++ b/third_party/BUILD
@@ -18,13 +18,13 @@ package(default_visibility = ["//:__subpackages__"])

 alias(
     name = "libssl",
-    actual = "@boringssl//:ssl",
+    actual = "@openssl//:ssl",
     tags = ["manual"],
 )

 alias(
     name = "libcrypto",
-    actual = "@boringssl//:crypto",
+    actual = "@openssl//:crypto",
     tags = ["manual"],
 )
