diff --git a/native/src/seal/serializable.h b/native/src/seal/serializable.h
index a940190..e490b30 100644
--- a/native/src/seal/serializable.h
+++ b/native/src/seal/serializable.h
@@ -135,6 +135,9 @@ namespace seal
             return obj_.save(out, size, compr_mode);
         }

+        const T& obj() const { return obj_; }
+
+        T& obj() { return obj_; }
     private:
         Serializable(T &&obj) : obj_(std::move(obj))
         {}

diff --git a/native/src/seal/context.cpp b/native/src/seal/context.cpp
index 887a1312..932d9774 100644
--- a/native/src/seal/context.cpp
+++ b/native/src/seal/context.cpp
@@ -477,7 +477,8 @@ namespace seal
         // more than one modulus in coeff_modulus. This is equivalent to expanding
         // the chain by one step. Otherwise, we set first_parms_id_ to equal
         // key_parms_id_.
-        if (!context_data_map_.at(key_parms_id_)->qualifiers_.parameters_set() || parms.coeff_modulus().size() == 1)
+        if (!context_data_map_.at(key_parms_id_)->qualifiers_.parameters_set() || parms.coeff_modulus().size() == 1 ||
+            !parms.use_special_prime())
         {
             first_parms_id_ = key_parms_id_;
         }
diff --git a/native/src/seal/encryptionparams.cpp b/native/src/seal/encryptionparams.cpp
index 31e07441..c34d0a45 100644
--- a/native/src/seal/encryptionparams.cpp
+++ b/native/src/seal/encryptionparams.cpp
@@ -23,8 +23,10 @@ namespace seal
             uint64_t poly_modulus_degree64 = static_cast<uint64_t>(poly_modulus_degree_);
             uint64_t coeff_modulus_size64 = static_cast<uint64_t>(coeff_modulus_.size());
             uint8_t scheme = static_cast<uint8_t>(scheme_);
+			 uint8_t use_special_prime = static_cast<uint8_t>(use_special_prime);
 
             stream.write(reinterpret_cast<const char *>(&scheme), sizeof(uint8_t));
+            stream.write(reinterpret_cast<const char *>(&use_special_prime), sizeof(uint8_t));
             stream.write(reinterpret_cast<const char *>(&poly_modulus_degree64), sizeof(uint64_t));
             stream.write(reinterpret_cast<const char *>(&coeff_modulus_size64), sizeof(uint64_t));
             for (const auto &mod : coeff_modulus_)
@@ -63,6 +65,10 @@ namespace seal
             // This constructor will throw if scheme is invalid
             EncryptionParameters parms(scheme);
 
+			 uint8_t use_special_prime;
+            stream.read(reinterpret_cast<char *>(&use_special_prime), sizeof(uint8_t));
+			 parms.set_use_special_prime(use_special_prime);
+
             // Read the poly_modulus_degree
             uint64_t poly_modulus_degree64 = 0;
             stream.read(reinterpret_cast<char *>(&poly_modulus_degree64), sizeof(uint64_t));
@@ -128,7 +134,8 @@ namespace seal
         size_t total_uint64_count = add_safe(
             size_t(1), // scheme
             size_t(1), // poly_modulus_degree
-            coeff_modulus_size, plain_modulus_.uint64_count());
+			 size_t(1), // use_special_prime
+            coeff_modulus_size);
 
         auto param_data(allocate_uint(total_uint64_count, pool_));
         uint64_t *param_data_ptr = param_data.get();
@@ -139,13 +146,15 @@ namespace seal
         // Write the poly_modulus_degree. Note that it will always be positive.
         *param_data_ptr++ = static_cast<uint64_t>(poly_modulus_degree_);
 
+		*param_data_ptr++ = static_cast<uint64_t>(use_special_prime_);
         for (const auto &mod : coeff_modulus_)
         {
             *param_data_ptr++ = mod.value();
         }
 
-        set_uint(plain_modulus_.data(), plain_modulus_.uint64_count(), param_data_ptr);
-        param_data_ptr += plain_modulus_.uint64_count();
+        // NOTE(juhou): we skip the plain modulus for parms_id
+        // set_uint(plain_modulus_.data(), plain_modulus_.uint64_count(), param_data_ptr);
+        // param_data_ptr += plain_modulus_.uint64_count();
 
         HashFunction::hash(param_data.get(), total_uint64_count, parms_id_);
 
diff --git a/native/src/seal/encryptionparams.h b/native/src/seal/encryptionparams.h
index 9e1fbe48..eb71c4ac 100644
--- a/native/src/seal/encryptionparams.h
+++ b/native/src/seal/encryptionparams.h
@@ -266,6 +266,11 @@ namespace seal
             random_generator_ = std::move(random_generator);
         }
 
+        inline void set_use_special_prime(bool flag)
+        {
+            use_special_prime_ = flag;
+        }
+
         /**
         Returns the encryption scheme type.
         */
@@ -274,6 +279,11 @@ namespace seal
             return scheme_;
         }
 
+        bool use_special_prime() const noexcept
+        {
+            return use_special_prime_;
+        }
+
         /**
         Returns the degree of the polynomial modulus parameter.
         */
@@ -501,6 +511,8 @@ namespace seal
 
         Modulus plain_modulus_{};
 
+		 bool use_special_prime_ = true;
+
         parms_id_type parms_id_ = parms_id_zero;
     };
 } // namespace seal
